service/storage_service.cc

```diff
@@ -5948,6 +5965,22 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     }
 
                     co_await ss._shared_token_metadata.stale_versions_in_use();
+
+                    auto&& tmeta = ss._shared_token_metadata.get()->tablets();
+                    for (auto&& [table_id, tmap] : tmeta.all_tables()) {
+                        for (auto&& [tid, trinfo] : tmap.transitions()) {
+                            if (trinfo.session_id) {
+                                if (trinfo.stage == locator::tablet_transition_stage::write_both_read_new) {
+                                    close_session(session_id(trinfo.session_id));
+                                } else {
+                                    create_session(session_id(trinfo.session_id));
+                                }
+                            }
+                        }
+                    }
+
+                    co_await drain_closing_sessions();
```

Comment:
How are we protected against drain_closing_sessions being executed concurrently and/or out of order? We must assume that two coordinators can be acting concurrently against every node.

Reply:
The intended post-condition of drain_closing_sessions() is that sessions which were closed prior to the call are drained after the call. 

Out of order execution is fine. The per-session close status and drain status is monotonic. Once session is closed it stays closed, and once it's drained it stays drained. Session stays around after it's closed and until drained, so later calls to drain_closing_sessions() will see sessions which are closed but not drained and call drain() on them. An absent session is both closed and drained. You can call session::drain() concurrently, and get the same post-condition (that it is drained).

I think there is a bug in drain_closing_sessions() where the iterator can be invalidated on concurrent call, I need to fix it by serializing the drain internally.

Reply:
What you describe is safe, but apparently not live. E.g. if there is an outdated coordinator running a request concurrently, or a leader change storm the effect will be just as gleb described - constant closing of sessions and no streaming progress made.

Reply:
I don't see why we would have constant closing of sessions and no liveness. As long as we're in the "streaming" stage in group0, session remains open. Session is only closed after it is nulled in group0. Coordinator failover doesn't close it. The new coordinator will retry streaming under the same session.

If the streaming process is still alive (`storage_service::_tablet_ops` keeps a registry), retry will not start a new one but join with the currently running streaming.

What's the scenario with outdated leader running a request concurrently? What kind of request? Executing a barrier should not disrupt sessions, because barrier itself only makes the replica reflect what is in group0, and what is in group0 is what the current coordinator intends. If the session is open in group0, barrier from the old coordinator will not close it.

Reply:
> I don't see why we would have constant closing of sessions and no liveness. As long as we're in the "streaming" stage in group0, session remains open. Session is only closed after it is nulled in group0. Coordinator failover doesn't close it. The new coordinator will retry streaming under the same session.
> 
> If the streaming process is still alive (`storage_service::_tablet_ops` keeps a registry), retry will not start a new one but join with the currently running streaming.
> 
> What's the scenario with outdated leader running a request concurrently? What kind of request? Executing a barrier should not disrupt sessions, because barrier itself only makes the replica reflect what is in group0, and what is in group0 is what the current coordinator intends. If the session is open in group0, barrier from the old coordinator will not close it.

I assume the outdated leader will invoke the very same RPC. In some cases it could also be the previous or the next RPC of the topology change. You discussion about safety/liveness above pertains to a concrete case you have in mind - but you add a generic instrument, and at least by its definition it's not standing for a single case, it's a general purpose mechanism to manage streaming. So I am trying to define the general safety guidelines for using it, not just for the case of the current tablet state machine.
For example, I can imagine that under some circumstances a deposed leader may decide to abort streaming and thus close the sessions. In this case it will perhaps be able to close them because RPC is also guarded by term. But it indicates that the session framework itself is not conforming to the open/closed principle, i.e. it relies on an  assumption that lays outside the session API.

Reply:
Can out of order management of the session map happen? E.g. imagine there is a coordinator A that requests creation of session S1 via a state machine command. Then there is a coordinator change and S1 is closed. We search the session list and don't find S1 because it was not added there yet. Looks like  it's best to avoid any interleaved operations with the session map, but it's not yet clear how and whether this in itself will not lead to any deadlocks. I'd still say that if session ids are switched to timeuuids, and the erasure from the map is done based on "delete everything with timestamp < given timeuuid", rather than delete individual sessions by id, the whole system will be more robust (although perhaps not as flexible).

Reply:
> Can out of order management of the session map happen? E.g. imagine there is a coordinator A that requests creation of session S1 via a state machine command. Then there is a coordinator change and S1 is closed. We search the session list and don't find S1 because it was not added there yet. Looks like it's best to avoid any interleaved operations with the session map, but it's not yet clear how and whether this in itself will not lead to any deadlocks. I'd still say that if session ids are switched to timeuuids, and the erasure from the map is done based on "delete everything with timestamp < given timeuuid", rather than delete individual sessions by id, the whole system will be more robust (although perhaps not as flexible).

Such problems cannot happen. Even though session management is invoked from the RPC, actual information does not come from the RPC, it comes from local group0 state machine. The meaning of RPC is "make sure sessions reflect group0". The order of such RPCs cannot affect the end result. So there is actually no concurrency regarding session changes. Local group0 follows the global group0 order, and local state of sessions follows the local group0 state, so it also follows the global group0 order of events.

Because group0 determines the state of sessions, we don't need session ids to be ordered.